network/p2p/cache/node_blocklist_wrapper.go

```diff
@@ -0,0 +1,191 @@
+// of the identities (whenever necessary) and modifies the `Ejected` flag only in
+// the copy.
+// The `NodeBlocklistWrapper` internally represents the `blocklist` as a map, to enable
+// performant lookup. However, the exported API works with `flow.IdentifierList` for
+// blocklist, as this is a broadly supported data structure which lends itself better
+// to config or command-line inputs.
+type NodeBlocklistWrapper struct {
+	m  sync.RWMutex
+	db *badger.DB
+
+	identityProvider id.IdentityProvider
+	blocklist        IdentifierSet // `IdentifierSet` is a map, hence efficient O(1) lookup
+}
+
+var _ id.IdentityProvider = (*NodeBlocklistWrapper)(nil)
+
+// NewNodeBlocklistWrapper wraps the given `IdentityProvider`. The blocklist is
+// loaded from the database (or assumed to be empty if no database entry is present).
+func NewNodeBlocklistWrapper(identityProvider id.IdentityProvider, db *badger.DB) (*NodeBlocklistWrapper, error) {
+	blocklist, err := retrieveBlocklist(db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read set of blocked node IDs from data base: %w", err)
+	}
+
+	return &NodeBlocklistWrapper{
+		db:               db,
+		identityProvider: identityProvider,
+		blocklist:        blocklist,
+	}, nil
+}
+
+// Update sets the wrapper's internal set of blocked nodes to `blocklist`. Empty list and `nil`
+// (equivalent to empty list) are accepted inputs. To avoid legacy entries in the data base, this
+// function purges the entire data base entry if `blocklist` is empty.
+// This implementation is _eventually consistent_, where changes are written to the data base first
+// and then (non-atomically!) the in-memory set of blocked nodes is updated. This strongly
+// benefits performance and modularity. No errors are expected during normal operations.
+func (w *NodeBlocklistWrapper) Update(blocklist flow.IdentifierList) error {
+	b := blocklist.Lookup() // convert slice to
+	err := persistBlocklist(b, w.db)
+	if err != nil {
+		return fmt.Errorf("failed to persist set of blocked nodes to the data base: %w", err)
+	}
+
+	w.m.Lock()
+	w.blocklist = b
+	w.m.Unlock()
+	return nil
+}
+
+// ClearBlocklist purges the set of blocked node IDs. Convenience function
+// equivalent to w.Update(nil). No errors are expected during normal operations.
+func (w *NodeBlocklistWrapper) ClearBlocklist() error {
+	return w.Update(nil)
+}
+
+// GetBlocklist returns the set of blocked node IDs.
+func (w *NodeBlocklistWrapper) GetBlocklist() flow.IdentifierList {
+	w.m.RLock()
+	defer w.m.RUnlock()
+
+	identifiers := make(flow.IdentifierList, 0, len(w.blocklist))
+	for i := range w.blocklist {
+		identifiers = append(identifiers, i)
+	}
+	return identifiers
+}
+
+// Identities returns the full identities of _all_ nodes currently known to the
+// protocol that pass the provided filter. Caution, this includes ejected nodes.
+// Please check the `Ejected` flag in the returned identities (or provide a
```

Comment:
Should we make it explicit by calling this method `IdentitiesWithEjected` and add another method `IdentitiesWithoutEjected`?

Reply:
The sole purpose of the `NodeBlocklistWrapper` is to implement the [`IdentityProvider` interface](https://github.com/onflow/flow-go/blob/7baa3f8b09ff8161349fba2f32d7fe61c6944d94/module/id/id_provider.go#L18-L40), which only has the method https://github.com/onflow/flow-go/blob/7baa3f8b09ff8161349fba2f32d7fe61c6944d94/module/id/id_provider.go#L23. 

Happy to add an additional method (like `UnmodifiedIdentities`) which would be outside of the [`IdentityProvider` interface](https://github.com/onflow/flow-go/blob/7baa3f8b09ff8161349fba2f32d7fe61c6944d94/module/id/id_provider.go#L18-L40), but not sure if there is use for it. We can always add it later. Do you have any use case in mind in relation to the admin command or so?

Reply:
I see. Looks good. Let's keep as is then

Reply:
revision: https://github.com/onflow/flow-go/pull/3344/commits/bd14684c0656b18aa4e22365d200bb94a7229fb7